### PR TITLE
Mirror 9 years of internal changes outward.

### PR DESCRIPTION
**Pull request by cpovirk**

### CODE REVIEW

**Comment by mwilli20** [comment]
@cpovirk Any hold up to merging this? This seems like a pretty nice thing to get on master.

**Comment by cpovirk** [comment]
Hi,

There have been a few things I'm working through with the actual owners of the library. We're making progress, but we're not quite there yet.

**Comment by mwilli20** [comment]
> Hi,
> 
> There have been a few things I'm working through with the actual owners of the library. We're making progress, but we're not quite there yet.

cool cool, just wanted to make sure this didn't fade away.

**Comment by jeffreypicard** [comment]
I would also very much like to see this merged. I'm using S2 heavily day-to-day and am currently building a lot of stuff like polygon indexes from scratch and I think the updates in here could be helpful, especially the addition of serializability since I do a lot of work in spark and have already written a bunch of kryo serializers, but I can't implement a proper partitioner in spark without them being serializable. I've hacked around that for the time being but want to wait for this update before throwing them into production.

Also might be worth noting to get this to build I needed an updated guava jar and the error_prone_annotations jar, neither of which are in the lib folder, the guava version being too old and the error_prone_annotations being non-existent.

**Comment by animesh-sharama** [comment]
Interested in seeing this getting merged, till the time will play with the cpovirk:moesync fork.

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,5 @@
-
+# .gitignore file for the open-source s2-geometry-java repo.
 build/
+target/
+.project
+.classpath
```
Comment (jmr) [inline .gitignore:5]: No newline at end of file.

**Comment by AlbertWigmore** [comment]
Are there any updates regarding this PR?

**Comment by aababilov** [comment]
I would also very much like to see this merged. This is important for https://github.com/MobilityData/gtfs-validator used at Google.

**Comment by aababilov** [comment]
That is awesome, thanks a lot!

How can we also update the Maven repository to provide the newer library version? Here I see that the latest 1.0.1 version was created in 2019: https://mvnrepository.com/artifact/io.sgr/s2-geometry-library-java

**Comment by cpovirk** [comment]
We have several threads going about what to do about a binary release. Most of them are on Google-internal mailing lists, so I'll give the summary here:

`io.sgr:s2-geometry-library-java` is a copy of the library published by [some other people](https://github.com/io-sgr/s2-geometry-library-java) under [the same Java package but a different Maven ID](https://jlbp.dev/JLBP-5).

The official artifact was [`com.google.common.geometry:s2-geometry`](https://mvnrepository.com/artifact/com.google.common.geometry/s2-geometry/1.0). That disappeared when Bintray (an alternative to Maven Central) [shut down](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/).

It sounds likely that we'll publish version 1.0 of `com.google.common.geometry:s2-geometry` again, this time on Maven Central. Ideally we'll also follow that with a 2.0. But I don't want to commit anyone else to that work right now, nor can I commit to it personally.

**Pull request merged**